### PR TITLE
Ana/4295 verify and store user accounts in db

### LINKED ISSUE #4295: Store user accounts in DB
Moving this from the archived lunie-api to be able to comment, reference, etc.

Resolve an OAuth user in the resolver.js and store it in the db. The frontend will add a header to send the uid of the user. The authentication remains completely with firebase.

In the resolver we will need to get the user from firebase. The FE will send an ID token in the header.
Verify Token
https://firebase.google.com/docs/auth/admin/verify-id-tokens#web
https://firebase.google.com/docs/auth/admin/custom-claims

DB
Create Schema (EMails remain in firebase)

user {
  uid
  email
  premium
  createdAt
  lastActive
}
On every verified token in resolver.js check if the user exists and either create the user or update lastActive.

### PR DESCRIPTION
**Pull request by Bitcoinera**
Closes #4295 

**Description:**

Working. Just needs the update part. I didn't finish it off because I wanted to confirm: `premium` would be a parameter coming in the query from the FE, right? This is assuming we will have some place in FE where people will be able to sign up for premium.

EDIT: Finally I added the premium thing as it seems quite sensible to me.

I have only one concern left: the `GOOGLE_APPLICATION_CREDENTIALS` variable seems only to be detected when doing `export GOOGLE_APPLICATION_CREDENTIALS`. So no matter what I do, in the `.env` file is always ignored. From a developer point of view, this is really annoying. All the time you need to remember this before running the API locally.

Any ideas here what I could be missing? All the other env variables are there but this one concretely is resisting

Thank you! 🚀

---

For contributor:

- [ ] Added changes entries. Run `yarn changelog` for a guided process.
- [ ] Reviewed `Files changed` in the github PR explorer
- [ ] Attach screenshots of the UI components on the PR description (if applicable)
- [ ] Scope of work approved for big PRs

For reviewer:

- [ ] Manually tested the changes on the UI

### CODE REVIEW

File: api/lib/resolvers.js [hunk 1]
```diff
@@ -294,6 +300,41 @@ const transactionMetadata = async (
   }
 }
 
+const storeUser = async (_, { idToken, premium }) => {
```
Comment (faboweb) [inline api/lib/resolvers.js:303]: let's move to a file like `authentication.js`

File: api/lib/resolvers.js [hunk 2]
```diff
@@ -294,6 +300,41 @@ const transactionMetadata = async (
...
+    // get user creation date
+    const userRecord = await firebaseAdmin.auth().getUser(decodedToken.uid)
+    // check if user already exists in DB
+    const storedUser = await database(config)('').getUser(decodedToken.uid)
+    // check if user already has premium as a custom claim
+    if (!userRecord.customClaims) {
+      // set premium field
+      await firebaseAdmin
+        .auth()
+        .setCustomUserClaims(decodedToken.uid, { premium: premium || false })
```
Comment (faboweb) [inline api/lib/resolvers.js:315]: what is the reason to use firebase claims?

File: api/lib/resolvers.js [hunk 3]
```diff
@@ -294,6 +300,41 @@ const transactionMetadata = async (
...
+    }
+    const user = {
+      uid: decodedToken.uid,
+      email: decodedToken.email,
+      premium: premium || false,
+      createdAt: userRecord.metadata.creationTime,
```
Comment (faboweb) [inline api/lib/resolvers.js:321]: as we have `createdAt` and `lastActive` in firebase, let's not store them in our db

File: api/lib/resolvers.js [hunk 4]
```diff
@@ -294,6 +300,41 @@ const transactionMetadata = async (
...
+    // check if user already has premium as a custom claim
+    if (!userRecord.customClaims) {
+      // set premium field
+      await firebaseAdmin
+        .auth()
+        .setCustomUserClaims(decodedToken.uid, { premium: premium || false })
+    }
+    const user = {
+      uid: decodedToken.uid,
+      email: decodedToken.email,
```
Comment (faboweb) [inline api/lib/resolvers.js:319]: we don't want to store the email for privacy reasons (for now)

File: api/lib/subscription/slashing.js [hunk 5]
```diff
@@ -26,18 +26,18 @@ class SlashingMonitor {
...
-        const missedBlocks = response.events['liveness.address'].map(
-          (address, index) => ({
-            validator: address,
-            missedBlocks: response.events['liveness.missed_blocks'][index]
-          })
-        )
-        console.log('Missed block', missedBlocks)
-      }
-    )
+    // this.client.subscribe(
```
Comment (faboweb) [inline api/lib/subscription/slashing.js:29]: uncomment?

**Comment by faboweb** [comment]
🚀 Deployed on https://5eea9d0711a0a985e47ca0c4--lunieio.netlify.app

File: api/lib/resolvers.js (hunk 2 above)
Comment (Bitcoinera) [reply api/lib/resolvers.js]: we need it to store the premium field inside the user. You put the docs reference in the issue description

File: api/lib/resolvers.js [hunk 6]
```diff
@@ -584,8 +573,10 @@ const resolvers = {
       if (!remoteFetch(dataSources, networkId).getAddressRole) return undefined
 
       return await remoteFetch(dataSources, networkId).getAddressRole(address)
-    },
-    storeUser: storeUser
+    }
+  },
+  Mutation: {
+    registerUser: registerUser
```
Comment (Bitcoinera) [inline api/lib/resolvers.js:467]: what is the advantage of setting this as a mutation?

File: api/lib/resolvers.js [hunk 7]
```diff
@@ -300,28 +301,16 @@ const transactionMetadata = async (
...
-    const userRecord = await firebaseAdmin.auth().getUser(decodedToken.uid)
     // check if user already exists in DB
     const storedUser = await database(config)('').getUser(decodedToken.uid)
-    // check if user already has premium as a custom claim
-    if (!userRecord.customClaims) {
-      // set premium field
-      await firebaseAdmin
-        .auth()
-        .setCustomUserClaims(decodedToken.uid, { premium: premium || false })
-    }
```
Comment (Bitcoinera) [inline api/lib/resolvers.js]: we are going to need including this at some point. Why not do it now and just set it always to false?

File: api/lib/resolvers.js (hunk 3 above)
Comment (Bitcoinera) [reply api/lib/resolvers.js]: it was in the issue description 🤷

File: api/lib/resolvers.js (hunk 2 above)
Comment (Bitcoinera) [reply api/lib/resolvers.js]: https://github.com/luniehq/lunie-api/issues/833

File: api/lib/resolvers.js [hunk 8]
```diff
@@ -18,9 +21,13 @@ const database = require('./database')
 const { getNotifications } = require('./notifications')
 const config = require('../config.js')
 const { logOverview } = require('./statistics')
-const BigNumber = require('bignumber.js')
 const networks = require('../data/networks')
 
+const firebaseServiceAccount = require('../firebaseCredentials.json')
+firebaseAdmin.initializeApp({
+  credential: firebaseAdmin.credential.cert(firebaseServiceAccount)
+})
```
Comment (Bitcoinera) [inline api/lib/resolvers.js:29]: Nice!

File: api/lib/resolvers.js (hunk 6 above)
Comment (Bitcoinera) [reply api/lib/resolvers.js:467]: over a Query I mean

File: app/config.js [hunk 9]
```diff
@@ -25,4 +25,6 @@ export default {
   referralLinks: {
     Coinbase: "https://coinbase-consumer.sjv.io/31vxX",
   },
+
+  firebaseConfig: JSON.parse(process.env.VUE_APP_FIREBASE_CONFIG || "{}"),
```
Comment (Bitcoinera) [inline app/config.js:29]: how would this look like in the `.env` file? I'm having trouble to test it

File: app/changes/ana_4295-verify-and-store-user-accounts-in-db [hunk 10]
```diff
@@ -0,0 +1 @@
+[Added] Register users in DB @faboweb
```
Comment (Bitcoinera) [inline app/changes/ana_4295-verify-and-store-user-accounts-in-db:1]: usurper!

File: app/config.js (hunk 9 above)
Comment (Bitcoinera) [reply app/config.js:29]: I'm getting all the time "unexpected end of JSON"

File: api/lib/resolvers.js (hunk 7 above)
Comment (Bitcoinera) [reply api/lib/resolvers.js]: I mean, we need this so the firebase user object contains this piece of information, so it cna propagate and be shared both by BE and FE

**Comment by Bitcoinera** [comment]
Tested and works. Would approve but would like this to get included: https://github.com/luniehq/lunie/pull/4301#discussion_r442156298

File: api/lib/resolvers.js (hunk 6 above)
Comment (faboweb) [reply api/lib/resolvers.js:467]: just conceptional: a query delivers data, a mutation manipulates data

**Comment by Bitcoinera** [comment]
Good to go!

**Comment by Bitcoinera** [comment]
Let's merge!! 💃

File: api/lib/resolvers.js (hunk 6 above)
Comment (Bitcoinera) [reply api/lib/resolvers.js:479]: Oh I see

**Comment by Bitcoinera** [comment]
🤔 Don't know why E2E logs aren't complete

**Pull request merged**